README.md
© 2018 OpenSwitch project. All information is contributed to and made available by OPX under the Creative Commons Attribution 4.0 International License (available at http://creativecommons.org/licenses/by/4.0/).

# opx-logging
This repository contains logging utilities for the OPX project which provide a simple wrapper for syslog activities. These wrapper functions and macros add the line number of the calling code, as well as other additional pieces of information.

You can optionally use the CLI for run-time modification of logging levels called _opx_logging_cli_.

All headers are located in OPX, and will be installed in the _*/usr/include/opx*_ folder (based on your prefix).

## Packages
- libopx-logging1\_*version*\_*arch*.deb — Utility libraries  
- libopx-logging-dev\_*version*\_*arch*.deb — Exported header files
- python-opx-logging\_*version*\_*arch*.deb — Python bindings

See [Architecture](https://github.com/open-switch/opx-docs/wiki/Architecture) for more information on the Logging module.

(c) 2018 Dell Inc. or its subsidiaries. All Rights Reserved.


debian/changelog
opx-logging (2.1.1+opx2) unstable; urgency=medium

  * Update: Copyright information

 -- Dell EMC <ops-dev@lists.openswitch.net>  Mon, 10 Dec 2018 14:16:24 -0700

opx-logging (2.1.1+opx1) unstable; urgency=medium

  * Update: Add new log levels and log output

 -- Dell EMC <ops-dev@lists.openswitch.net>  Tue, 06 Nov 2018 14:16:24 -0700

opx-logging (2.1.1) unstable; urgency=medium

  * Update: Modification in print_log_config function in opx_logging_cli script

 -- Dell EMC <ops-dev@lists.openswitch.net>  Tue, 13 Feb 2018 14:16:24 -0700

opx-logging (2.1.0) unstable; urgency=medium

  * Update: Added a new api to reload the config for processes which already handles the SIGHUP
  * Update: Added prototype of event_log_reload_config() in event_log.h

 -- Dell EMC <ops-dev@lists.openswitch.net>  Thu, 25 Dec 2017 14:16:24 -0700

opx-logging (2.0.0+opx1) unstable; urgency=medium

  * Bugfix: Modifications to print_log_config function in opx_logging_cli script to make the command
            "opx_logging_cli show" work

 -- Dell EMC <ops-dev@lists.openswitch.net>  Thu, 22 Dec 2017 13:16:24 -0700

opx-logging (2.0.0) unstable; urgency=medium

  * Update: Designed logging module to use systemd journal
  * Update: Modifications to opx_logging_cli script and logs up to and including INFO are disabled by default
  * Update: Support for SIGHUP for log file updates and enable all

 -- Dell EMC <ops-dev@lists.openswitch.net>  Mon, 19 June 2017 13:16:24 -0700

opx-logging (1.0.1) unstable; urgency=medium

  * Initial release.

 -- Dell EMC <ops-dev@lists.openswitch.net>  Wed, 01 Feb 2017 09:44:15 -0800


src/event_log.cpp
/*
 * Copyright (c) 2018 Dell Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License"); you may
 * not use this file except in compliance with the License. You may obtain
 * a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
 *
 * THIS CODE IS PROVIDED ON AN *AS IS* BASIS, WITHOUT WARRANTIES OR
 * CONDITIONS OF ANY KIND, EITHER EXPRESS OR IMPLIED, INCLUDING WITHOUT
 * LIMITATION ANY IMPLIED WARRANTIES OR CONDITIONS OF TITLE, FITNESS
 * FOR A PARTICULAR PURPOSE, MERCHANTABLITY OR NON-INFRINGEMENT.
 *
 * See the Apache Version 2.0 License for specific language governing
 * permissions and limitations under the License.
 */

/*
 * filename: event_log.cpp
 */

#include "event_log.h"
#include "event_log_types.h"

#include <stdarg.h>
#include <stdio.h>
#include <stdlib.h>
#include <pthread.h>
#include <string.h>
#include <unistd.h>

#include <mutex>
#include <type_traits>
#include <unordered_map>
#include <string>
#include <sstream>
#include <vector>
#include <algorithm>
#include <systemd/sd-journal.h>

/*Future home can be determined but we have a general default and can be overridden by environmental variables*/
/*maybe take this path from the make environment in the near future*/
#define _LOG_CONFIG_FILE "/etc/opx/evlog.cfg"

/****
 * Environment variable for getting the vendor and OS string
 * OPX_LOGPREFIX is the name of the environment variable that
 * stores the brand specific log prefix
 **/
#define LOG_PREFIX "OPX_LOGPREFIX"

/* Log format literals */
#define SP_TAG " "
#define COMMA_TAG ","
#define ALARM_TAG "[alarm]"
#define EVENT_TAG "[event]"
#define DEBUG_TAG "[debug]"
#define AUDIT_TAG "[audit]"
#define CHASSIS_TAG "Node.1-Unit.1:PRI"
#define JOURNAL_MSG_TAG "MESSAGE="
static std::string log_prefix;
static const char *_log_file_name = _LOG_CONFIG_FILE;
static pthread_once_t event_log_lib_inited = PTHREAD_ONCE_INIT;

static std::unordered_map<std::string,std::vector<bool>> * _mods;
static std::recursive_mutex *_mutex;

static size_t _min_level;
static size_t _max_level;

static time_t _last_signal;
static time_t _last_loaded=-1;

static bool _enable_all=true;

template <typename T>
inline void __fix_level(T &level) {
    if (level <_min_level|| level>=_max_level) {
        level = _max_level;
    }
}

inline void __fix_module(const char* &mod) {
    if (mod==nullptr) mod = "";
}


static void _reload_config(int sig) {
    time(&_last_signal);
}


void event_log_reload_config(){
    _reload_config(0);
}

#include <sys/signal.h>

/****
 * @brief Fetches the environment variable by variable name
 * @param var environment variable name in context
 * @return value stored in the env variable as a standard string
 * In case brand environment variable is not present, the
 * default return value of this function is an empty string
 **/
static std::string get_environment_variable(const char *var)
{
    char *prefix = getenv(var);

    /****
     * Initializing string constructor with a nullptr is an undefined
     * behavior, which could be disastrous. In this code, I have taken
     * care to fall back to defaults in-case vendor environment is not
     * setup
     **/

    return prefix != NULL ? std::string{prefix} : std::string{""};
}

void event_log_init() {
    /****
     * openlogs API is used to open syslog sockect with special option
     * to record PID. call to closelog API is optional.
     * Ref: https://www.gnu.org/software/libc/manual/html_node/openlog.html
     **/
    openlog (NULL, LOG_PID, DEFAULT_TRACE_FACILITY);
    log_prefix = get_environment_variable(LOG_PREFIX);

    syslog(LOG_MAKEPRI(DEFAULT_TRACE_FACILITY,LOG_DEBUG), "LOG PREFIX: %s\n",
                (log_prefix.size())?log_prefix.c_str():"getenv returned NULL");

    std::vector<size_t> _levels = { EV_T_LOG_EMERG,
                                    EV_T_LOG_ALERT,
                                    EV_T_LOG_CRIT,
                                    EV_T_LOG_ERR,
                                    EV_T_LOG_WARNING,
                                    EV_T_LOG_NOTICE,
                                    EV_T_LOG_INFO,
                                    EV_T_LOG_DEBUG
                                 };

    std::sort(_levels.begin(),_levels.end());
    _min_level = _levels.front();
    _max_level = _levels.back();

    _mods = new std::remove_reference<decltype(*_mods)>::type();
    _mutex = new std::recursive_mutex;

    struct sigaction _a;
    memset(&_a,0,sizeof(_a));
    _a.sa_handler = &_reload_config;

    //cam't do much if it fails but really there should be no valid failure case
    ///TODO Need to remove SIGUSR1 in a follow up update - temporary until SIGHUP totally integrated
    (void)sigaction(SIGUSR1,&_a,nullptr);

    (void)sigaction(SIGHUP,&_a,nullptr);

    //support Systemd environment override
    const char * _file_name = getenv("EV_LOGGING_CFG");
    if (_file_name!=nullptr) _log_file_name = _file_name;
}

inline void evt_check_log_init(){
    pthread_once(&event_log_lib_inited,event_log_init);
}

static std::vector<bool> &__get_mask(const char *mod) {
    std::lock_guard<std::recursive_mutex> lg(*_mutex);
    auto it = _mods->find(mod);
    if (it==_mods->end()) {
        std::vector<bool> &_flags = (*_mods)[mod];
        _flags.resize(_max_level+1);
        for ( auto  it : _flags ) {
            it = true;
        }
        _flags[EV_T_LOG_INFO] = false;
        _flags[EV_T_LOG_DEBUG] = false;
        return _flags;
    }
    return it->second;
}

void event_logging_set_enabled_flag(const char *mod,size_t loglvl, bool how) {
    evt_check_log_init();
    std::lock_guard<std::recursive_mutex> lg(*_mutex);
    __fix_level(loglvl);
    __fix_module(mod);

    std::vector<bool> &_flags = __get_mask(mod);
    _flags[loglvl] = how;
}

static inline const char * __adjust_mod_name(const char *mod) {
    const char *_tmpl = "ev_log_t_";
    while (*mod==*_tmpl) { ++mod,++_tmpl; }
    return mod;
}

 bool event_logging_is_enabled(const char *mod,size_t loglvl) {
    evt_check_log_init();

    std::lock_guard<std::recursive_mutex> lg(*_mutex);

    mod = __adjust_mod_name(mod);
    if (_last_loaded!=_last_signal) {
        _last_loaded = _last_signal;
        FILE *_fp = fopen(_log_file_name,"r");
        if (_fp!=nullptr) {
            _mods->clear();
            char _buff[256];
            while (fgets(_buff,sizeof(_buff)-1,_fp)!=nullptr) {
                char *_sep = strchr(_buff,' ');
                if (_sep==nullptr) continue;
                if (_sep==_buff) continue; //empty/invalid
                if (strlen(_buff)==0) continue; //invalid/empty

                *_sep = '\0';
                if (*(++_sep)==0) continue;

                //Overall log enable/disable
                if (strcmp(_buff,"ALL")==0) {
                    _enable_all = atoi(_sep);
                    continue;
                }

                bool _valid_line = false;

                //determine logging level
                size_t _level = atoi(_sep);
                bool _enable = false;

                //determine enable/disable
                _sep = strchr(_sep,' ');
                if (_sep!=nullptr ) {
                    ++_sep;
                    if (*_sep!=0) {
                        _enable = atoi(_sep);
                        _valid_line = true;
                    }
                }

                if (!_valid_line) continue;

                __fix_level(_level);
                event_logging_set_enabled_flag(_buff,_level,_enable);
            }
            fclose(_fp);
        }
    }

    if (!_enable_all) return false;

    __fix_module(mod);
    __fix_level(loglvl);


    /* Check if the sublevel and module id are in the valid enum range
     * loglvl is guaranteed to be the safe since user can not control its value
     */

    std::vector<bool> &_flags = __get_mask(mod);
    return _flags[loglvl];
}

bool event_log_msg(const char *module, int level, int ident, const char *file,
            const char * ln, const char *fun, const char *tag, int err, const char *msg,...) {
    va_list lst;
    va_start(lst, msg);

    evt_check_log_init();

    std::stringstream aug_msg;
    uint32_t facility = DEFAULT_TRACE_FACILITY;
    switch(ident) {
        case ev_log_trace:
            vsyslog(LOG_MAKEPRI(level, facility), msg, lst);
            break;
        case ev_log_debug:
            facility = LOG_LOCAL4;
            aug_msg << CHASSIS_TAG
                    << SP_TAG << DEBUG_TAG << COMMA_TAG
                    << SP_TAG << msg;

            vsyslog(LOG_MAKEPRI(level, facility), aug_msg.str().c_str(), lst);
            break;
        case ev_log_audit:
            facility = EV_T_LOG_AUDIT;
            aug_msg << CHASSIS_TAG
                    << SP_TAG << AUDIT_TAG << COMMA_TAG
                    << SP_TAG << msg;

            vsyslog(LOG_MAKEPRI(level, facility), aug_msg.str().c_str(), lst);
            break;
        case ev_log_event:
            facility = LOG_LOCAL4;
            aug_msg << CHASSIS_TAG
                    << SP_TAG << EVENT_TAG << COMMA_TAG
                    << SP_TAG;

            if (!log_prefix.empty())
                aug_msg << log_prefix << SP_TAG;

            aug_msg << msg;
            vsyslog(LOG_MAKEPRI(level, facility), aug_msg.str().c_str(), lst);
            break;
        case ev_log_alarm:
        {
            /****
             * The following code was added to allow alarm manager retireve events pertaining to
             * alarm event. After design discussions, it was  decided that event_log_msg will be
             * morphed to  suit params required for ALM. Therefore, the original  param list may
             * be used in different context in the following code.
             */
            va_list args_cpy;
            va_copy (args_cpy, lst);
            const auto sz = std::vsnprintf (nullptr, 0, msg, lst)+1; //Extra byte allocated

            try {
                std::string detail_desc (sz, '\0');
                std::vsnprintf (&detail_desc.front(), sz, msg, args_cpy);
                va_end (args_cpy);

                aug_msg << JOURNAL_MSG_TAG << CHASSIS_TAG
                        << SP_TAG << ALARM_TAG << COMMA_TAG
                        << SP_TAG;
                if (!log_prefix.empty())
                    aug_msg << log_prefix << SP_TAG;
                aug_msg << tag;

                sd_journal_send (
                        "PRIORITY=%i", LOG_ALERT,               /* alarm entries have higher precedence */
                        "TAG=alarm",                            /* Filtering tag */
                        aug_msg.str().c_str(), module,
                        fun, detail_desc.c_str(),               /* Log Message */
                        "EV_NAME=%s", module,                   /* Name of the alarm event */
                        "EV_DESC=%s", fun,                      /* Description from alarm event */
                        "EV_SEV=%i", level,                     /* Alarm event severity */
                        "EV_ID=%i", err,                        /* Alarm event id */
                        "EV_SRC=%s", file,                      /* Alarm event source */
                        "EV_ACTION=%s", ln,                     /* Alarm event action */
                        "EV_DET_DESC=%s", detail_desc.c_str(),  /* Detail desciption from user */
                        "SYSLOG_FACILITY=%i", 20,               /* alarm event facility */
                        NULL);
            } catch (const std::bad_alloc&) {
                va_end (args_cpy);
                va_end(lst);
                return false;
            }
            break;
        }
        default:
            sd_journal_printv_with_location(level, file,ln,fun, msg,lst);
    }

    va_end(lst);
    return true;
}



inc/opx/event_log.h
/*
 * Copyright (c) 2018 Dell Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License"); you may
 * not use this file except in compliance with the License. You may obtain
 * a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
 *
 * THIS CODE IS PROVIDED ON AN *AS IS* BASIS, WITHOUT WARRANTIES OR
 * CONDITIONS OF ANY KIND, EITHER EXPRESS OR IMPLIED, INCLUDING WITHOUT
 * LIMITATION ANY IMPLIED WARRANTIES OR CONDITIONS OF TITLE, FITNESS
 * FOR A PARTICULAR PURPOSE, MERCHANTABLITY OR NON-INFRINGEMENT.
 *
 * See the Apache Version 2.0 License for specific language governing
 * permissions and limitations under the License.
 */

/*
 * filename: event_log.h
 */

#ifndef __EVENT_LOG_H
#define __EVENT_LOG_H

#include "event_log_types.h"

#include <errno.h>
#include <stdbool.h>
#include <stddef.h>



#ifdef __cplusplus
extern "C" {
#endif


/*Convert the module name to a string*/
#define _EV_LOG_TO_STRING(x) #x

/**
 * @brief log the msg
 *
 * @param MOD a module from event_log_types.h
 * @param LVL a syslog level to be used for logging from event_log_types.h
 * @param ID  a event id (private id for trace logs)
 * @param msg the message and parameters will follow this data
 */
#define EV_LOGGING(MOD,LVL,ID,msg, ...) \
    (void)((event_logging_is_enabled(_EV_LOG_TO_STRING(MOD), EV_T_LOG_ ## LVL)) && \
        ((EV_T_LOG_ ## LVL != EV_T_LOG_DEBUG) ?    \
            event_log_msg(_EV_LOG_TO_STRING(MOD),EV_T_LOG_ ## LVL, 0, "CODE_FILE=" __FILE__,\
                "CODE_LINE=" _EV_LOG_TO_STRING(__LINE__),__func__,ID,0,"[%s:%s], " msg,_EV_LOG_TO_STRING(MOD),ID,##__VA_ARGS__):    \
            event_log_msg(_EV_LOG_TO_STRING(MOD),EV_T_LOG_ ## LVL, 0, "CODE_FILE=" __FILE__,\
                "CODE_LINE=" _EV_LOG_TO_STRING(__LINE__),__func__,ID,0,"[%s:%s] %s:%s:%d, " msg,_EV_LOG_TO_STRING(MOD),ID,\
                __FILE__,__func__,__LINE__,##__VA_ARGS__)))

/**
 * @brief change all logging state (enable or disable)
 *
 * @param state is to enable (true) or disable (false) all logging state
 *
 */
void event_log_change_all_state(bool state);

/**
 * @brief checks if logging is enabled for a given module and log-level
 *
 * @param mod is a string tag for the module
 * @param loglvl is the numeric log-level from event_log_types.h
 * @return current logging status for given parameters (true/false)
 */
bool event_logging_is_enabled(const char *mod,size_t loglvl);


/**
 * @brief     Set the state of the module's log level.  By default, all trace logs
 *             are disabled
 * @param mod is a string tag for the module
 * @param loglvl is the numeric log-level from event_log_types.h
 * @return current logging status for given parameters (true/false)
 */
void event_logging_set_enabled_flag(const char *mod,size_t loglvl, bool how);

/**
 * @brief The following function will send the log to the appropriate backend
 *         logging system.  This is not meant to be directly called by applications
 *         the logging macros should be used as a front end
 *
 * @param module is a string tag
 * @param level is the numeric log-level from event_log_types.h
 * @param ident is a numeric id if provided
 * @param file is the filename
 * @param ln is the line number of the entry
 * @param fun is the function call name
 * @param tag is a string that can identify the issue or otherwise qualify the log message
 * @param err is the error number
 * @param msg is the message which needs to be logged
 *
 */
bool event_log_msg(const char *module, int level, int ident, const char *file,
        const char * ln, const char *fun, const char *tag, int err, const char *msg,...);

/**
 * @brief - load the latest logging configuration after it has been changed
 *
 * Loading of the latest logging configuration is already handled when user
 * send the SIGUP signal to the process. This API can be used if some third
 * party applications are handling the SIGHUP signal and in this case new logging
 * configuration needs to be loaded at run time.
 *
 */

void event_log_reload_config(void);

#include "event_log_dep.h"

#ifdef __cplusplus
}
#endif

#endif


inc/opx/event_log_types.h
/*
 * Copyright (c) 2018 Dell Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License"); you may
 * not use this file except in compliance with the License. You may obtain
 * a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
 *
 * THIS CODE IS PROVIDED ON AN *AS IS* BASIS, WITHOUT WARRANTIES OR
 * CONDITIONS OF ANY KIND, EITHER EXPRESS OR IMPLIED, INCLUDING WITHOUT
 * LIMITATION ANY IMPLIED WARRANTIES OR CONDITIONS OF TITLE, FITNESS
 * FOR A PARTICULAR PURPOSE, MERCHANTABLITY OR NON-INFRINGEMENT.
 *
 * See the Apache Version 2.0 License for specific language governing
 * permissions and limitations under the License.
 */

/*
 * filename: event_log_types.h
 */

#ifndef __EVENT_LOG_TYPES_H
#define __EVENT_LOG_TYPES_H

#include <syslog.h>

#define EV_T_LOG_EMERG LOG_EMERG
#define EV_T_LOG_ALERT LOG_ALERT
#define EV_T_LOG_CRIT LOG_CRIT
#define EV_T_LOG_ERR LOG_ERR
#define EV_T_LOG_WARNING LOG_WARNING
#define EV_T_LOG_NOTICE LOG_NOTICE
#define EV_T_LOG_INFO LOG_INFO
#define EV_T_LOG_DEBUG LOG_DEBUG
#define EV_T_LOG_TRACE LOG_DEBUG

#define DEFAULT_LOG_LEVEL EV_T_LOG_NOTICE

//  Used in the script for automation if new log level added please update it accordingly.
#define MAX_LOG_LEVEL 7

//  Used to identify special  logs
enum ev_log_e {
    ev_log_trace,
    ev_log_alarm,
    ev_log_debug,
    ev_log_event,
    ev_log_audit
};

#
/* Audit log facility */
#define EV_T_LOG_AUDIT (13<<3)

/* Default trace facility */
#define DEFAULT_TRACE_FACILITY LOG_LOCAL7

//   Enums are used as index, so be careful when adding a new sublevel, only increment by one
enum e_log_sublevel {
    ev_log_s_CRITICAL=0,
    ev_log_s_MAJOR=1,
    ev_log_s_MINOR=2,
    ev_log_s_WARNING=3,
    ev_log_s_MAX,    //add all items above this line
};

#include "event_log_dep.h"

#endif


src/python_extension/event_log.c
/*
 * Copyright (c) 2018 Dell Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License"); you may
 * not use this file except in compliance with the License. You may obtain
 * a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
 *
 * THIS CODE IS PROVIDED ON AN *AS IS* BASIS, WITHOUT WARRANTIES OR
 * CONDITIONS OF ANY KIND, EITHER EXPRESS OR IMPLIED, INCLUDING WITHOUT
 * LIMITATION ANY IMPLIED WARRANTIES OR CONDITIONS OF TITLE, FITNESS
 * FOR A PARTICULAR PURPOSE, MERCHANTABLITY OR NON-INFRINGEMENT.
 *
 * See the Apache Version 2.0 License for specific language governing
 * permissions and limitations under the License.
 */

#include <python2.7/Python.h>

#include "event_log.h"

extern int event_log_name_to_mod(const char *name, int len);

static PyObject *__all_logs(int log_level, PyObject *self, PyObject *args)
{
    int        lvl, lineno;
    const char *modname, *tag, *filename, *func, *msg;

    if (!PyArg_ParseTuple(args, "sisssis", &modname, &lvl, &tag, &filename, &func, &lineno, &msg))  return (NULL);

    if (event_logging_is_enabled(modname, lvl)) {
        char _ln[100];
        char _file[512];
        snprintf(_ln,sizeof(_ln)-1,"CODE_LINE=%d",(int)lineno);
        snprintf(_file,sizeof(_file)-1,"CODE_FILE=%s",filename);
        event_log_msg(modname, lvl, 0,_file,_ln,func,tag,0,msg);
    }

    return (Py_BuildValue(""));
}


static PyObject *
log_trace(PyObject *self, PyObject *args)
{
    return __all_logs(EV_T_LOG_DEBUG,self,args);
}

static PyObject *
log_info(PyObject *self, PyObject *args)
{
    return __all_logs(EV_T_LOG_INFO,self,args);
}

static PyObject *
log_err(PyObject *self, PyObject *args)
{
    return __all_logs(EV_T_LOG_ERR,self,args);
}

static PyObject * logging(PyObject *self, PyObject *args)
{
    return __all_logs(EV_T_LOG_INFO,self,args);
}

static PyMethodDef methods[] = {
    { "log_trace", log_trace, METH_VARARGS, "Log trace" },
    { "log_info",  log_info,  METH_VARARGS, "Log info" },
    { "log_err",   log_err,   METH_VARARGS, "Log error" },
    { "logging",   logging,   METH_VARARGS, "Log msg" },
    { NULL, NULL, 0, NULL }        /* Sentinel */
};

PyMODINIT_FUNC
initevent_log(void)
{
    PyObject *m = Py_InitModule("event_log", methods);

    PyModule_AddIntConstant(m, "EMERG", EV_T_LOG_EMERG);
    PyModule_AddIntConstant(m, "ALERT", EV_T_LOG_ALERT);
    PyModule_AddIntConstant(m, "CRIT", EV_T_LOG_CRIT);
    PyModule_AddIntConstant(m, "ERR", EV_T_LOG_ERR);
    PyModule_AddIntConstant(m, "WARNING", EV_T_LOG_WARNING);
    PyModule_AddIntConstant(m, "NOTICE", EV_T_LOG_NOTICE);
    PyModule_AddIntConstant(m, "INFO", EV_T_LOG_INFO);
    PyModule_AddIntConstant(m, "DEBUG", EV_T_LOG_DEBUG);}


inc/opx/event_log_dep.h
/*
 * Copyright (c) 2018 Dell Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License"); you may
 * not use this file except in compliance with the License. You may obtain
 * a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
 *
 * THIS CODE IS PROVIDED ON AN *AS IS* BASIS, WITHOUT WARRANTIES OR
 * CONDITIONS OF ANY KIND, EITHER EXPRESS OR IMPLIED, INCLUDING WITHOUT
 * LIMITATION ANY IMPLIED WARRANTIES OR CONDITIONS OF TITLE, FITNESS
 * FOR A PARTICULAR PURPOSE, MERCHANTABLITY OR NON-INFRINGEMENT.
 *
 * See the Apache Version 2.0 License for specific language governing
 * permissions and limitations under the License.
 */

#ifndef LOGGING_INC_EVENT_LOG_DEP_H_
#define LOGGING_INC_EVENT_LOG_DEP_H_


/**
 * @deprecated THIS API HAS BEEN DEPRECATED, PLEASE DO NOT USE THIS
 *
 * @note Please use @sa EV_LOGGING API instead
 *
 * @brief trace a message to through syslog
 *
 * @param MOD a module from event_log_types.h
 * @param a logging level (0-3)
 * @param a event id (private id for trace logs
 * @param message a message (printf formatted.. no return char please)
 * @param parameters for the message in standard printf format
 */

#define EV_LOG_TRACE(MOD, LVL, ID, message, ... ) \
        EV_LOGGING(MOD,DEBUG,ID,message,  ## __VA_ARGS__)

/**
 * @deprecated THIS API HAS BEEN DEPRECATED, PLEASE DO NOT USE THIS
 *
 * @note Please use @sa EV_LOGGING API instead
 *
 * @brief log an error event.  may generate traps or the like...
 *
 * @param MOD a module from event_log_types.h
 * @param a event id (private id for trace logs
 * @param message a message (printf formatted.. no return char please)
 * @param parameters for the message in standard printf format
 */
#define EV_LOG_ERR(MOD, LVL, ID, message, ... ) \
    EV_LOGGING(MOD,ERR,ID,message, ##__VA_ARGS__)

/**
 * @deprecated THIS API HAS BEEN DEPRECATED, PLEASE DO NOT USE THIS
 *
 * @note Please use @sa EV_LOGGING API instead
 *
 * @brief info message
 *
 * @param MOD a module from event_log_types.h
 * @param a event id (private id for trace logs
 * @param message a message (printf formatted.. no return char please)
 * @param parameters for the message in standard printf format
 */
#define EV_LOG_INFO(MOD, LVL, ID, message, ... ) \
    EV_LOGGING(MOD,INFO,ID,message, ##__VA_ARGS__)

/**
 * @deprecated THIS API HAS BEEN DEPRECATED, PLEASE DO NOT USE THIS
 *
 * @note Please use @sa EV_LOGGING API instead
 *
 * @brief This is a wrapper for all of the above macros that will simplify the calling
 * @param what can be INFO, TRACE, ERR, DEBUG, CON_TRACE
 * @param MOD is the module concatenated with ev_log_t_ to create the real subsystem
 * @param LVL is a numeric level from 0-100 that indicates severity.  For error logs
 *             not applied at this point
 * @param ID a tag that is added to the log (can be used as classifier)
 * @param msg the message and parameters will follow this data
 */
#define EV_LOG( what, MOD, LVL, ID, msg, ...) \
    EV_LOGGING(MOD,what,ID,msg, ##__VA_ARGS__)

/**
 * @deprecated THIS API HAS BEEN DEPRECATED, PLEASE DO NOT USE THIS
 *
 * @note Please use @sa EV_LOGGING API instead
 *
 * @brief log the errno to the error log
 *
 * @param MOD a module from event_log_types.h
 * @param a event id (private id for trace logs
 * @param ERRNO is the errno that you want printed to the log
 */
#define EV_LOG_ERRNO(MOD, LVL, ID, ERRNO) \
    EV_LOGGING(MOD,ERR,ID,"Error occured %d",ERRNO)


/**
 * @deprecated THIS list of logging levels has been depreciated - do not use
 *
*/
//  Enums are used as index, so be careful when adding a new module, only increment by one
enum e_event_log_types_enums {
    ev_log_t_NULL=0,
    ev_log_t_NPU=1,
    ev_log_t_BOARD=2,
    ev_log_t_SYSTEM=3,
    ev_log_t_QOS=4,
    ev_log_t_COM=5,
    ev_log_t_INTERFACE=6,
    ev_log_t_NETLINK=7,
    ev_log_t_ROUTE=8,
    ev_log_t_ACL=9,
    ev_log_t_SFP=10,
    ev_log_t_MGMT=11,
    ev_log_t_DSAPI=12,
    ev_log_t_DB_SQL=13,
    ev_log_t_NDI=14,
    ev_log_t_SAI_FDB=15,
    ev_log_t_SAI_VLAN=16,
    ev_log_t_SAI_PORT=17,
    ev_log_t_SAI_SWITCH=18,
    ev_log_t_SAI_ROUTER=19,
    ev_log_t_SAI_ROUTER_INTF=20,
    ev_log_t_SAI_NEIGHBOR=21,
    ev_log_t_SAI_NEXT_HOP=22,
    ev_log_t_SAI_NEXT_HOP_GROUP=23,
    ev_log_t_SAI_ROUTE=24,
    ev_log_t_SAI_MIRROR=25,
    ev_log_t_NAS_L2=26,
    ev_log_t_NAS_COM=27,
    ev_log_t_FWD_SVCS_POLICY=28,
    ev_log_t_FWD_SVCS_QOS=29,
    ev_log_t_FWD_SVCS_NETMON=30,
    ev_log_t_SAI_LAG=31,
    ev_log_t_LACP=32,
    ev_log_t_XSTP=33,
    ev_log_t_LLDP=34,
    ev_log_t_DOT1X=35,
    ev_log_t_L2MAC=36,
    ev_log_t_L2CMN=37,
    ev_log_t_NAS_OS=38,
    ev_log_t_SAI_STP=39,
    ev_log_t_BGP=40,
    ev_log_t_OSPFV2=41,
    ev_log_t_OSPFV3=42,
    ev_log_t_RTM=43,
    ev_log_t_L3CMN=44,
    ev_log_t_PALD=45,
    ev_log_t_PAS_MONITOR=46,
    ev_log_t_PAS=47,
    ev_log_t_PAS_FUSE=48,
    ev_log_t_PAS_TIMER=49,
    ev_log_t_SAI_SAMPLEPACKET=50,
    ev_log_t_ENV_TMPCTL=51,
    ev_log_t_SWUPDATE=52,
    ev_log_t_INFRA_ALM=53,
    ev_log_t_INFRA_CHM=54,
    ev_log_t_INFRA_ETL=55,
    ev_log_t_INFRA_IFM=56,
    ev_log_t_INFRA_ISSU=57,
    ev_log_t_INFRA_PM=58,
    ev_log_t_SAI_HOSTIF=59,
    ev_log_t_SAI_QUEUE=60,
    ev_log_t_SAI_MAPS=61,
    ev_log_t_SAI_POLICER=62,
    ev_log_t_SAI_WRED=63,
    ev_log_t_SAI_SCHEDULER=64,
    ev_log_t_SAI_SCHEDULER_GRP=65,
    ev_log_t_SAI_QOS=66,
    ev_log_t_SERV_HW=67,
    ev_log_t_DATASTORE=68,
    ev_log_t_L3_SERVICES=69,
    ev_log_t_IPM=70,
    ev_log_t_MGMT_INTF=71,
    ev_log_t_L2_SERVICES=72,
    ev_log_t_FWD_SVCS_SFLOW=73,
    ev_log_t_INFRA_PPM=74,
    ev_log_t_SVC_SUPPORTASSIST=75,
    ev_log_t_INFRA_NDM=76,
    ev_log_t_INFRA_EQM=77,
    ev_log_t_SAI_UDF=78,
    ev_log_t_SAI_BUFFERS=79,
    ev_log_t_SWITCH_RES_MGR=80,
    ev_log_t_SERV_SW=81,
    ev_log_t_INFRA_NDS=82,
    ev_log_t_APP_VLT=83,
    ev_log_t_SUPPORT_BUNDLE=84,
    ev_log_t_APP_OPENFLOW=85,
    ev_log_t_SAI_HASH=86,
    ev_log_t_AFS_UTILS=87,
    ev_log_t_VRRP=88,
    ev_log_t_OTM=89,
    ev_log_t_INFRA_AFS=90,
    ev_log_t_INFRA_COMMON=91,
    ev_log_t_SAI_TUNNEL=92,
    ev_log_t_SERV_TRIGG=93,
    ev_log_t_APP_LADF=94,
    ev_log_t_ISCSI_OPT=95,
    ev_log_t_DCBX=96,
    ev_log_t_NBR_MGR=97,
    ev_log_t_FABRIC_AGENT=98,
    ev_log_t_DNV_APPS=99,
    ev_log_t_APP_FC=100,
    ev_log_t_SAI_FC=101,
    ev_log_t_MAX                ///add all items above this line
};



#endif /* LOGGING_INC_EVENT_LOG_DEP_H_ */
